packages/react/toast/src/Toast.tsx

```diff
@@ -233,20 +271,67 @@ const ToastViewport = React.forwardRef<ToastViewportElement, ToastViewportProps>
         tabIndex={-1}
         // incase list has size when empty (e.g. padding), we remove pointer events so
         // it doesn't prevent interactions with page elements that it overlays
-        style={{ pointerEvents: context.toastCount > 0 ? undefined : 'none' }}
+        style={{ pointerEvents: hasToasts ? undefined : 'none' }}
       >
+        {hasToasts && (
+          <FocusProxy
+            ref={headFocusProxyRef}
+            onEntranceFocus={() => focusFirst(getSortedCandidates({ direction: 'backwards' }))}
+          />
+        )}
         {/**
          * tabindex on the the list so that it can be focused when items are removed. we focus
          * the list instead of the viewport so it announces number of items remaining.
          */}
-        <Primitive.ol tabIndex={-1} {...viewportProps} ref={composedRefs} />
+        <Collection.Slot scope={__scopeToast}>
+          <Primitive.ol tabIndex={-1} {...viewportProps} ref={composedRefs} />
+        </Collection.Slot>
+        {hasToasts && (
+          <FocusProxy
+            ref={tailFocusProxyRef}
+            onEntranceFocus={() => focusFirst(getSortedCandidates({ direction: 'forwards' }))}
+          />
+        )}
       </DismissableLayer.Branch>
     );
   }
 );
 
 ToastViewport.displayName = VIEWPORT_NAME;
 
+/* -----------------------------------------------------------------------------------------------*/
+
+const FOCUS_PROXY_NAME = 'ToastFocusProxy';
+
+type FocusProxyElement = React.ElementRef<typeof VisuallyHidden>;
+type VisuallyHiddenProps = Radix.ComponentPropsWithoutRef<typeof VisuallyHidden>;
+interface FocusProxyProps extends VisuallyHiddenProps {
+  onEntranceFocus(): void;
+}
+
+const FocusProxy = React.forwardRef<FocusProxyElement, ScopedProps<FocusProxyProps>>(
+  (props, forwardedRef) => {
+    const { __scopeToast, onEntranceFocus, ...proxyProps } = props;
+    const context = useToastProviderContext(FOCUS_PROXY_NAME, __scopeToast);
+
+    return (
+      <VisuallyHidden
+        aria-hidden
+        tabIndex={0}
```

Comment:
should these have similar styles to focus guards to ensure no involuntary page scrolls (`position: fixed` primarily)?

Reply:
Good spot!